src/codegen/query.cpp

```diff
@@ -96,46 +176,95 @@ void Query::Execute(std::unique_ptr<executor::ExecutorContext> executor_context,
+  LOG_INFO("Using codegen interpreter to execute plan");
+
+  // Timer
+  Timer<std::ratio<1, 1000>> timer;
+  if (stats != nullptr) {
+    timer.Start();
+  }
+
+  // Create Bytecode
+  interpreter::BytecodeFunction init_bytecode =
+      interpreter::BytecodeBuilder::CreateBytecodeFunction(
+          code_context_, llvm_functions_.init_func);
+  interpreter::BytecodeFunction plan_bytecode =
+      interpreter::BytecodeBuilder::CreateBytecodeFunction(
+          code_context_, llvm_functions_.plan_func);
+  interpreter::BytecodeFunction tear_down_bytecode =
+      interpreter::BytecodeBuilder::CreateBytecodeFunction(
+          code_context_, llvm_functions_.tear_down_func);
 
-  // Compile the code
-  if (!code_context_.Compile()) {
-    return false;
+  // Time initialization
+  if (stats != nullptr) {
+    timer.Stop();
+    stats->interpreter_prepare_ms = timer.GetDuration();
+    timer.Reset();
+    timer.Start();
   }
 
-  LOG_TRACE("Setting up Query ...");
+  // Call init
+  LOG_TRACE("Calling query's init() ...");
+  try {
+    interpreter::BytecodeInterpreter::ExecuteFunction(
+        init_bytecode, reinterpret_cast<char *>(function_arguments));
+  } catch (...) {
+    interpreter::BytecodeInterpreter::ExecuteFunction(
+        tear_down_bytecode, reinterpret_cast<char *>(function_arguments));
+    throw;
+  }
 
-  // Get pointers to the JITed functions
-  init_func_ = (compiled_function_t)code_context_.GetRawFunctionPointer(
-      query_funcs.init_func);
-  PELOTON_ASSERT(init_func_ != nullptr);
+  if (stats != nullptr) {
+    timer.Stop();
+    stats->init_ms = timer.GetDuration();
+    timer.Reset();
+    timer.Start();
+  }
 
-  plan_func_ = (compiled_function_t)code_context_.GetRawFunctionPointer(
-      query_funcs.plan_func);
-  PELOTON_ASSERT(plan_func_ != nullptr);
+  // Execute the query!
+  LOG_TRACE("Calling query's plan() ...");
+  try {
+    interpreter::BytecodeInterpreter::ExecuteFunction(
+        plan_bytecode, reinterpret_cast<char *>(function_arguments));
+  } catch (...) {
+    interpreter::BytecodeInterpreter::ExecuteFunction(
+        tear_down_bytecode, reinterpret_cast<char *>(function_arguments));
+    throw;
+  }
+
+  // Timer plan execution
+  if (stats != nullptr) {
+    timer.Stop();
+    stats->plan_ms = timer.GetDuration();
+    timer.Reset();
+    timer.Start();
+  }
 
-  tear_down_func_ = (compiled_function_t)code_context_.GetRawFunctionPointer(
-      query_funcs.tear_down_func);
-  PELOTON_ASSERT(tear_down_func_ != nullptr);
+  // Clean up
+  LOG_TRACE("Calling query's tearDown() ...");
+  interpreter::BytecodeInterpreter::ExecuteFunction(
+      tear_down_bytecode, reinterpret_cast<char *>(function_arguments));
 
-  LOG_TRACE("Query has been setup ...");
+  // No need to cleanup if we get an exception while cleaning up...
+  if (stats != nullptr) {
+    timer.Stop();
+    stats->tear_down_ms = timer.GetDuration();
+  }
 
-  // All is well
+  // TODO(marcel): return value
```

Comment:
What should this function return?

Reply:
The Execute functions return always true at the moment, which is not helpful. 
I will change them to return void, because an exception is thrown in an error case anyway. (either from inside the query or from the Interpreter, if the function cannot be translated)